Libraries/LibGUI/ModelIndex.h

```diff
@@ -68,6 +71,7 @@ class ModelIndex {
     int m_row { -1 };
     int m_column { -1 };
     void* m_internal_data { nullptr };
+    void* m_role_specific_data { nullptr };
```

Comment:
So what is this role-specific data, exactly? It's specific to *what* role, seems like it's only one member of data? And how's it different from the already existing internal data?

Reply:
it's specific to whatever role that was passed to `Model::data()`.
perhaps changing the signature of that function is the more sensible thing to do?

the internal data is...well, internal to the model that owns the index, this thing is a part of is public interface.

Reply:
Oh right, changing the signature of `Model::data` was too much of a chore, now I remember why I put that into the index lol

Reply:
Changing the signature of `data()` sounds way more sensible than putting this into the index.

Reply:
I don't understand why an additional argument is necessary in the first place.
Why does the underlying target model of a filter model need to know anything about the filtering state?

Reply:
Yeah, that, too. The model being filtered shouldn't know it's being filtered, it just provides the data.

Reply:
We have two options here
- ask the model "does the stuff at this index match this search term?"
- ask the model for the data and do a specific kind of test (e.g. `String.contains`)

I personally think the first makes much more sense.

Reply:
But doesn't that mean each and every model will have to implement searching?

Perhaps we need to scale the ambition back, and only implement this for the man pages model. It lets you get the file path, and then the outer thing could search in that. We could think of a generic search+filter functionality for models later.

Reply:
> But doesn't that mean each and every model will have to implement searching?

Yes, and that's clearly wrong. A filtering-proxy-model should be just that: a proxy. Something that sits in between a model and a view, filtering the target model, and requiring no additional capabilities from the target model.

Reply:
I really don't see the problem here, we can just fall back to checking the `Role::Display` string if the model doesn't implement searching (it's not really searching, it's a "does this _index_ match this term").
The only real cost is adding an argument to `data()` that will mostly do nothing at the moment.

Reply:
If you want an escape hatch to do optimized searches without going through generic data lookup, then we can add *that* instead of messing with the simplicity of `data()` :)

A model could have something like a `virtual TriState data_matches(ModelIndex, Variant) const;` that returns True/False if it is/isn't a match, and Unknown if it's not supported.

Reply:
_that_ works for me.